### PR TITLE
Memory usage improvements

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

ECC: make private key field 'k' able to be smaller when ALT_ECC_SIZE is
defined.
WOLFSSL_SMALL_STACK_CACHE: allocate temps using new macros.

When compiled for low memory usage (including ALT_ECC_SIZE) and RSA/DH included:
Previously:
```
  Application: client_p256_rsa_tls12
    Accept/Connect: 914
    Max in use SSL: 23010
    Errors: 0
  Application: server_p256_rsa_tls12
    Accept/Connect: 1458
    Max in use SSL: 31313
    Errors: 0
  Application: client_p256_rsa_tls12_nca
    Accept/Connect: 914
    Max in use SSL: 19549
    Errors: 0
  Application: server_p256_rsa_tls12_nca
    Accept/Connect: 1458
    Max in use SSL: 23764
    Errors: 0

  Application: client_p256_ecc_tls12
    Accept/Connect: 914
    Max in use SSL: 24690
    Errors: 0
  Application: server_p256_ecc_tls12
    Accept/Connect: 1458
    Max in use SSL: 27614
    Errors: 0
  Application: client_p256_ecc_tls12_nca
    Accept/Connect: 914
    Max in use SSL: 23282
    Errors: 0
  Application: server_p256_ecc_tls12_nca
    Accept/Connect: 1458
    Max in use SSL: 23300
    Errors: 0
```

With PR changes:
```
  Application: client_p256_rsa_tls12
    Accept/Connect: 914
    Max in use SSL: 17854
    Errors: 0
  Application: server_p256_rsa_tls12
    Accept/Connect: 1458
    Max in use SSL: 23705
    Errors: 0
  Application: client_p256_rsa_tls12_nca
    Accept/Connect: 914
    Max in use SSL: 14393
    Errors: 0
  Application: server_p256_rsa_tls12_nca
    Accept/Connect: 1458
    Max in use SSL: 20187
    Errors: 0

  Application: client_p256_ecc_tls12
    Accept/Connect: 914
    Max in use SSL: 13708
    Errors: 0
  Application: server_p256_ecc_tls12
    Accept/Connect: 1458
    Max in use SSL: 16629
    Errors: 0
  Application: client_p256_ecc_tls12_nca
    Accept/Connect: 914
    Max in use SSL: 12302
    Errors: 0
  Application: server_p256_ecc_tls12_nca
    Accept/Connect: 1458
    Max in use SSL: 13396
    Errors: 0
```

# Testing

Standard and memory usage.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

**Comment by SparkiDev** [comment]
retest this please

File: src/pk.c [hunk 1]
```diff
@@ -30,6 +30,14 @@
     #include <wolfssl/wolfcrypt/random.h>
 #endif
 
+#ifdef HAVE_ECC
+    #include <wolfssl/wolfcrypt/ecc.h>
+#endif
+#ifndef WOLFSSL_HAVE_ECC_KEY_GET_PRIV
```
Comment (dgarske) [inline src/pk.c:36]: This replicated section... is it to support use against older FIPS snapshots that won't have `wc_ecc_key_get_priv` and need the `&` deference?

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -5547,7 +5594,11 @@ static int _ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key,
         DECLARE_CURVE_SPECS(ECC_CURVE_FIELD_COUNT);
 
         /* setup the key variables */
-        err = mp_init(&key->k);
+#ifndef ALT_ECC_SIZE
+        err = mp_init(key->k);
+#else
+        alt_fp_init(key->k);
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:5600]: Should there be a return code check here in case init wasn't called and ->k isn't set to `->ka`?

**Review by dgarske — CHANGES_REQUESTED** [review]
Very good work on this! Just a couple minor things.

File: wolfssl/wolfcrypt/ecc.h [hunk 3]
```diff
@@ -347,14 +347,15 @@ typedef struct ecc_set_type {
  * The ALT_ECC_SIZE option only applies to stack based fast math USE_FAST_MATH.
  */
 
-#if !defined(USE_FAST_MATH) && !defined(WOLFSSL_SP_MATH_ALL) && \
-    !defined(WOLFSSL_SP_MATH)
-    #error USE_FAST_MATH must be defined to use ALT_ECC_SIZE
+#if defined(USE_INTEGER_HEAP_MATH)
+    #error Cannot use integr math with ALT_ECC_SIZE
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/ecc.h:351]: fixed

File: wolfcrypt/src/ecc.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c:5600]: Change to set key->k to be key->ka before call.

**Review by dgarske — APPROVED** [review]
Great work! Also confirmed with async / QuickAssist.

**Comment by dgarske** [comment]
[1Z26173W0194956631](https://track.aftership.com/ups/1Z26173W0194956631)

**Comment by dgarske** [comment]
Retest this please

**Review by dgarske — APPROVED** [review]

**Pull request merged**